### ISSUE TITLE
Change in resolveStaticField is not suitable for lazyloaded datatable

### ISSUE THREAD

**Issue by mertsincan** [primefaces/primefaces#5162]
Reported By PRO User;
> We have a datatable with lazyloading.
> We use code like
> filterBy="#{orderType}"
> because it does not matter, if property named "orderType" do not exist. It is just constant with which load method works.
> 
> In PF 7 it stopped working. I found this change
> https://github.com/primefaces/primefaces/issues/4185
> which cause #{orderType} to return null as name of filtered field.

Test code;
test.xhtml
```
<h:form id="form">

    <p:dataTable var="car" value="#{dtLazyView.lazyModel}" paginator="true" rows="10"
                 paginatorTemplate="{RowsPerPageDropdown} {FirstPageLink} {PreviousPageLink} {CurrentPageReport} {NextPageLink} {LastPageLink}"
                 rowsPerPageTemplate="5,10,15" selectionMode="single" selection="#{dtLazyView.selectedCar}" id="carTable" lazy="true">
        <p:column headerText="Id" sortBy="#{car.id}" filterBy="#{car.id}">
            <h:outputText value="#{car.id}" />
        </p:column>
        <p:column headerText="Year" sortBy="#{car.year}">
            <h:outputText value="#{car.year}" />
        </p:column>
        <p:column headerText="Year from" filterBy="#{anything.yearFrom}">
            <h:outputText value="#{car.year}" />
        </p:column>    
        <p:column headerText="Year to" filterBy="#{anything.yearTo}">
            <h:outputText value="#{car.year}" />
        </p:column>     
        <p:column headerText="Brand">
            <h:outputText value="#{car.brand}" />
        </p:column>
        <p:column headerText="Color">
            <h:outputText value="#{car.color}" />
        </p:column>
    </p:dataTable>

</h:form>
```
LazyCarDataModel.java
```
public class LazyCarDataModel extends LazyDataModel<Car> {

  private List<Car> datasource;

  public LazyCarDataModel(List<Car> datasource) {
    this.datasource = datasource;
  }

  @Override
  public Car getRowData(String rowKey) {
    for (Car car : datasource) {
      if (car.getId().equals(rowKey))
        return car;
    }

    return null;
  }

  @Override
  public Object getRowKey(Car car) {
    return car.getId();
  }

  @Override
  public List<Car> load(int first, int pageSize, String sortField, SortOrder sortOrder, Map<String, Object> filters) {
    List<Car> data = new ArrayList<Car>();

    int yearFrom = 0;
    int yearTo = 4000;

    if (filters.containsKey("yearFrom")) {
      String filterProperty = "yearFrom";
      Object filterValue = filters.get(filterProperty);

      if (filterValue != null) {
        yearFrom = Integer.valueOf(filterValue.toString());
      }
    }

    if (filters.containsKey("yearTo")) {
      String filterProperty = "yearTo";
      Object filterValue = filters.get(filterProperty);

      if (filterValue != null) {
        yearTo = Integer.valueOf(filterValue.toString());
      }
    }

    // filter
    for (Car car : datasource) {
      if (yearFrom <= car.getYear() && car.getYear() <= yearTo) {
        data.add(car);
      }
    }

    // sort
    if (sortField != null) {
      Collections.sort(data, new LazySorter(sortField, sortOrder));
    }

    // rowCount
    int dataSize = data.size();
    this.setRowCount(dataSize);

    // paginate
    if (dataSize > pageSize) {
      try {
        return data.subList(first, first + pageSize);
      } catch (IndexOutOfBoundsException e) {
        return data.subList(first, first + (dataSize % pageSize));
      }
    } else {
      return data;
    }
  }
}
```
The other codes; https://www.primefaces.org/showcase/ui/data/datatable/lazy.xhtml

**Comment by mertsincan** [comment]
Hi @Rapster,

I reverted this commit; https://github.com/primefaces/primefaces/commit/5f818a7238102facf62ca1ded88d1a7aceccf101

Users can add #{anything.field} to filterBy/sortBy to pass your expression. Please try the above test code.

Best Regards,

**Comment by Rapster** [comment]
I think `DataTable#resolveStaticField` works as expected, your PRO user is actually lucky it worked before. I don't think this method was meant to support "fake" field. There are 2 scenarios in this ticket:

1. PRO user mentionned:
> We use code like filterBy="#{orderType}" because it does not matter, if property named orderType" do not exist. It is just constant with which load method works.

If it's a constant then why not use `<f:attribute name="orderType" value="MARKET_ORDER" /`> and then in the load method retrieve it it with `Components.getCurrentComponent().get("orderType")`? Or alternative, instanciate the model with the `orderType` so you can reuse it in `LazyDataModel#load` method. IMO, option 2 is better

2. In your sample, it's a different story, I actually think the use case is wrong. Why not defined a custom filter in column "year" with `p:datePicker` range selection and do the filtering from here? Instead of having 2 "fakes" colums which does not even exist in the Car POJO

**Comment by melloware** [comment]
@mertsincan the build is broken because the unit test was failing.  I just submitted a PR with fixed unit test and I think I made a better fix to resolving the static field.

**Comment by Rapster** [comment]
I would prefer an answer from Mert before going any further. Unit test failing shows the method is not doing what it's suppose to do. It's fiddling what it's done here, to solve this use case: simply create two attributes in the Car POJO (to adapt with PRO user code of course)

**Comment by martin654** [comment]
Hi. 
I think that the assumption that filtering can happen only on Car POJO attributes is not correct. With lazy datatable we often use backend sorting, filtering and pagination. Filters values are sent to backend together with info about current page, page size, sorting column and sort order. In such a scenario the coupling between POJO attributes and filter values is very loose. It basically says that some filter value should be sent to backend. Backend can even do some business logic with those filters, it doesn't always have to simply filter value from table column.

**Comment by Rapster** [comment]
> Backend can even do some business logic with those filters, it doesn't always have to simply filter value from table column

Who said your view representation should be similar to your domain model/database? Anyway, I need to have a deeper look at this to fully understand how these filters works and also I'm not happy with using String manipulation as it is a "performance killer" and boilerplate code

**Comment by melloware** [comment]
@Rapster let me know what you come up with.  But I will say there is a few spots in the code where its looking for a string between two string such as between "{" and "}". We should move this method into LangUtils which does exactly that from Commons Lang.

```java
/**
     * <p>Gets the String that is nested in between two Strings.
     * Only the first match is returned.</p>
     *
     * <p>A {@code null} input String returns {@code null}.
     * A {@code null} open/close returns {@code null} (no match).
     * An empty ("") open and close returns an empty string.</p>
     *
     * <pre>
     * StringUtils.substringBetween("wx[b]yz", "[", "]") = "b"
     * StringUtils.substringBetween(null, *, *)          = null
     * StringUtils.substringBetween(*, null, *)          = null
     * StringUtils.substringBetween(*, *, null)          = null
     * StringUtils.substringBetween("", "", "")          = ""
     * StringUtils.substringBetween("", "", "]")         = null
     * StringUtils.substringBetween("", "[", "]")        = null
     * StringUtils.substringBetween("yabcz", "", "")     = ""
     * StringUtils.substringBetween("yabcz", "y", "z")   = "abc"
     * StringUtils.substringBetween("yabczyabcz", "y", "z")   = "abc"
     * </pre>
     *
     * @param str  the String containing the substring, may be null
     * @param open  the String before the substring, may be null
     * @param close  the String after the substring, may be null
     * @return the substring, {@code null} if no match
     * @since 2.0
     */
    public static String substringBetween(final String str, final String open, final String close) {
        if (str == null || open == null || close == null) {
            return null;
        }
        final int start = str.indexOf(open);
        if (start != INDEX_NOT_FOUND) {
            final int end = str.indexOf(close, start + open.length());
            if (end != INDEX_NOT_FOUND) {
                return str.substring(start + open.length(), end);
            }
        }
        return null;
    }
```

**Comment by Rapster** [comment]
Okay so, after reading the documentation my guess is that you could use [`Column#field`](https://www.primefaces.org/docs/vdl/7.0/primefaces-p/column.html#field):

> Name of the field to pass lazy load method for filtering and sorting. If not specified, filterBy-sortBy values are used to identify the field name.

@mertsincan @martin654 could you confirm it does the same as your tweak?

**Comment by Rapster** [comment]
@melloware I think this method could be useful yes, but not in this particularly case

**Comment by martin654** [comment]
@Rapster Hi, reproducer attached. "Column#filed" is not working for filtering nor sorting.

[primefaces-test-master_column_field.zip](https://github.com/primefaces/primefaces/files/3712214/primefaces-test-master_column_field.zip)

**Comment by Rapster** [comment]
I just tried your use case:

1. You passed incorrect value to field, pass a String value not an EL expression (unless it has a proper value)
2. [Column#field](https://www.primefaces.org/docs/vdl/7.0/primefaces-p/column.html#field) does exactly what you want
3. we'll have to revert that commit sooner or later since it has regression on #4185

**Comment by Rapster** [comment]
@martin654 could you please have a new try with my suggestion, reverting that commit has regression on #4185

**Comment by Rapster** [comment]
8.0 release is coming soon, and it'd be nice to hear from you since as I said it's causing a regression that need to be fixed

**Comment by martin654** [comment]
Hi, sorry for delay. I will try your suggestions and give feedback.

**Comment by martin654** [comment]
Hi, sorry, I had only short time for retest. With PF version 7.0.4 (same as in reproducer above) I can use field attribute as suggested:

```xml
<p:column headerText="Brand" field="brand" filterable="true" sortable="true">
    <f:facet name="filter">
        <p:selectOneMenu onchange="PF('_carTable').filter()" value="#{dtLazyView.brandFilterValue}" >
            <f:selectItem itemLabel="All" itemValue="#{null}" noSelectionOption="true"/>
            <f:selectItems value="#{dtLazyView.brandFilter}" var="item" itemValue="#{item}"/>
        </p:selectOneMenu>
    </f:facet>
    <h:outputText value="#{car.brand}" />
</p:column>
```

I would expect that filterable and sortable set to true would enable sorting and filtering, but the UI doesn't contain filtering inputs any more, nor the sorting feature.

It is true that the **field** is propagated into the Map<String, Object> filters in the load method.

Any ideas with the filtering inputs? 

[primefaces-test-master_column_field2.zip](https://github.com/primefaces/primefaces/files/3905401/primefaces-test-master_column_field2.zip)

**Comment by Rapster** [comment]
> It is true that the field is propagated into the Map<String, Object> filters in the load method.

Yes, it is true

> I would expect that filterable and sortable set to true would enable sorting and filtering, but the UI doesn't contain filtering inputs any more, nor the sorting feature.

I'm having a look

**Comment by Rapster** [comment]
The code is clear, in `DataTableRenderer#encodeColumnHeader`:

```java
ValueExpression columnSortByVE = column.getValueExpression(Column.PropertyKeys.sortBy.toString());
ValueExpression columnFilterByVE = column.getValueExpression(Column.PropertyKeys.filterBy.toString());
boolean sortable = (columnSortByVE != null && column.isSortable());
boolean filterable = (columnFilterByVE != null && column.isFilterable());
```

`UIColumn#isFilterable` and `UIColumn#isSortable` return true by default. So, yes to enable filters/sorters: sortBy/filterBy must be set.

But am I happy with this? No. Just like you, I would expect that, if `field` set and `sortable`/`filterable `enabled then filters/sorters should be rendered.

`sortBy`/`filterBy `should be used as fallback (custom sort/filter): i18n for example.

`DataTable#resolveColumnField` does a good job though: first `UIColumn#getField` is called and if not set, use sortBy/filterBy

You can create a ticket for this

**Comment by Rapster** [comment]
@martin654 I'll revert those changes soon, but I wanna be sure you're okay with it

**Comment by mertsincan** [comment]
Hi @Rapster,

I think the resolveStaticField works as expected. It doesn't need to check if the value of the filterBy/sortBy variable is at the server. Because, some Users can do something in the lazy-load method using these values. I think they should decide its value.  Exp; "#{bean.x.y}" -> in load method: x.y  
I think this is a more flexible use for users. 
> ... I don't think this method was meant to support "fake" field
- Whether the data/field is fake or not is something the user can control. My idea is that we should not intervene.

Best Regards,

**Comment by Rapster** [comment]
Okay let's start with the original problem:

> We use code like filterBy="#{orderType}" because it does not matter, if property named "orderType" do not exist. It is just constant with which load method works.

As I said, you're lucky it worked before because `DataTable#resolveStaticField` always expected an EL expression on this form: `#{var.property}`, but the code before 7.0 was unintentionnally okay with `#{property}`, as a result it causes issue #4185 (fixed in 7.0) with facelet tags (and they've been in JSF for a long time so we should support it).

Now, if it is a constant that you want, and I clearly understand what your client is trying to achieve then I already gave the solution to this problem, using  [`Column#field`](https://www.primefaces.org/docs/vdl/7.0/primefaces-p/column.html#field):

> Name of the field to pass lazy load method for filtering and sorting. If not specified, filterBy-sortBy values are used to identify the field name.

It literally does what your client is expecting. Based on your sample, it becomes:
```xml
<p:dataTable var="car" value="#{dtLazyView.lazyModel}" paginator="true" rows="10"
                         paginatorTemplate="{RowsPerPageDropdown} {FirstPageLink} {PreviousPageLink} {CurrentPageReport} {NextPageLink} {LastPageLink}"
                         rowsPerPageTemplate="5,10,15" selectionMode="single" selection="#{dtLazyView.selectedCar}" id="carTable" lazy="true">
    <p:column headerText="Id" sortBy="#{car.id}" filterBy="#{car.id}">
        <h:outputText value="#{car.id}" />
    </p:column>
    <p:column headerText="Year" sortBy="#{car.year}">
        <h:outputText value="#{car.year}" />
    </p:column>
    <p:column headerText="Year from" field="yearFrom" filterBy="#{car.year}">
        <h:outputText value="#{car.year}" />
    </p:column>
    <p:column headerText="Year to" field="yearTo" filterBy="#{car.year}">
        <h:outputText value="#{car.year}" />
    </p:column>
    <p:column headerText="Brand">
        <h:outputText value="#{car.brand}" />
    </p:column>
    <p:column headerText="Color">
        <h:outputText value="#{car.color}" />
    </p:column>
</p:dataTable>
```     

Backing bean becomes in PF 8.0:
```java
    @Override
    public List<Car> load(int first, int pageSize, String sortField, SortOrder sortOrder, List<FilterMeta> filters) {
        List<Car> data = new ArrayList<>();

        int yearFrom = 0;
        int yearTo = 4000;

        FilterMeta filter = filters.stream()
                .filter(f -> f.getFilterField().equalsIgnoreCase("yearFrom"))
                .findFirst()
                .orElse(null);
        if (filter != null) {
            Object filterValue = filter.getFilterValue();
            if (filterValue != null) {
                yearFrom = Integer.parseInt(filterValue.toString());
            }
        }

        filter = filters.stream()
                .filter(f -> f.getFilterField().equalsIgnoreCase("yearTo"))
                .findFirst()
                .orElse(null);
        if (filter != null) {
            Object filterValue = filter.getFilterValue();
            if (filterValue != null) {
                yearTo = Integer.parseInt(filterValue.toString());
            }
        }

        // filter
        for (Car car : datasource) {
            if (yearFrom <= car.getYear() && car.getYear() <= yearTo) {
                data.add(car);
            }
        }

        // sort
        if (sortField != null) {
            Collections.sort(data, new LazySorter(sortField, sortOrder));
        }

        // rowCount
        int dataSize = data.size();
        setRowCount(dataSize);

        // paginate
        if (dataSize > pageSize) {
            try {
                return data.subList(first, first + pageSize);
            } catch (IndexOutOfBoundsException e) {
                return data.subList(first, first + (dataSize % pageSize));
            }
        } else {
            return data;
        }
    }

    @Override
    public List<Car> load(int first, int pageSize, List<SortMeta> sortMeta, List<FilterMeta> filterMeta) {
        return Collections.emptyList();
    }
```

To sum up: `DataTable#resolveStaticField` was working perfectly fine in 7.0, and you have everything that you need to make your use case working in 7.0

**Comment by martin654** [comment]
Thank you both for your comments. 

Yes, the approach with Colum#field works as described. My concern is that I have to specify filterBy and sortBy expressions to enable filtering / sorting. I can fill anything I want to filterBy and sortBy, they just have to contain something...:

```
<p:column headerText="Year" field="year" filterBy="#{fooo.yyyyyy}" sortBy="#{foo.xxxxx}">
    <h:outputText value="#{car.year}" />
</p:column> 
```

I think that Column attributes filterable="true" sortable="true" should be used for this and filterBy and sortBy should be ommited at all.

@mertsincan ?

**Comment by Rapster** [comment]
Cool, I'm glad we clear the air. i'll revert the commit then, and close the other ticket also.

Regarding your last coment, I don't think we can change anything right now since 8.0 is getting release very soon. The impact of these changes would be way too big at this stage.

Though, I'm gonna create a ticket to throw few ideas on datatable, I think there are few things we can simplify.

**Comment by tandraschko** [comment]
@Rapster please don't revert/commit anything currently if it isn't a bugfix.
If it's just for better code or something, wait for 9.0

**Comment by Rapster** [comment]
not better code, the original commit is a bug fix of #4185 (fixed in 7.0)

**Comment by tandraschko** [comment]
Then please provide a PR - Mert can then decide if we merge it for 8.0 or later.
